tests/python_on_whales/components/test_container.py

```diff
@@ -702,3 +702,170 @@ def test_attach_sig_proxy_argument(run_mock: Mock, inspect_mock: Mock) -> None:
     run_mock.assert_called_once_with(
         docker.client_config.docker_cmd + ["attach", test_container_name], tty=True
     )
+
+
+@patch("python_on_whales.components.container.cli_wrapper.run")
```

Comment:
For the unit tests, I would prefer if we don't use `patch` unless it's absolutely necessary.  You can try for exemple to test the "never" option by removing the docker image first and do a docker run after, then check that an exception has been raised.

To check the "missing" you can test by using docker run when the docker image is not present locally, it should not crash. When the docker image is present locally, it should not try to get an updated version.

You can update a docker image in a registry by using the fixture that launch a registry locally. Then you'll have rights to pull and push to it. Take a look at the conftest.py

for the "always", we can check that it works by pulling the image first, pushing an updated version to the registry and then trying to do docker run again. It should have updated the docker image locally.

It's always better to avoid `patch` if we can, because then not all the code is executed in the unit test

Reply:
Actually, I was actually using `patch` on purpose so as to not execute all the code.

Since `python_on_whales` is a wrapper for the Docker CLI, why isn't it enough to just ensure that the CLI is called with the proper arguments for each case? Testing `python_on_whales` by actually calling Docker and analyzing its outcomes sounds to me like not only the wrapper is being tested but also Docker itself (up to some extent).

I think that makes sense for functions such as `copy_to` and `copy_from` that are not part of the CLI itself.
It may also make sense when testing the function `utils.run` since it calls the actual CLI.

However, in this case, I think it suffices to ensure that the expected arguments are passed to the CLI and that the images are pulled as expected (without having to actually pull them).

Let me know your thoughts on this :)

Reply:
Well in theory, yes, in practice, we might make an error in the expected command in the unit tests. Let's say that you make a unit test and you say with patch that you expect the command called to be 
`docker run -v /tmp:/tmp,/somedir:/some_other_dir ubuntu`. Then the test will pass because you thought this was the right syntax, but by running the real code, you would get an error and thus correct the wrong code. I had to do some trial and error when making this package because the docker CLI documentation is not always clear, so I would prefer that the real code is running to avoid this kind of issues